### ISSUE TITLE
Discussion on C++11 version

### ISSUE THREAD

**Issue by incebellipipo** [waterlinked/dvl-a50-ros-driver#9]
Hi,

TL;DR;
I've been using DVL A50 for sometime now. I had to chance to test both serial and ethernet communication. I've written a C++ version of the ROS driver. I've tested it with ROS Noetic. I'll be testing it for other versions of ROS. I'll be updating this thread and the repository as I test.

https://github.com/GSO-soslab/waterlinked_dvl_ros

- Python version differences in never distributions can cause incompatibility. E.g. newer might not have `update-alternatives` set up, therefore, there might not be a default python interpreter that it can find. https://github.com/waterlinked/dvl-a50-ros-driver/blob/4e8f887a061cdcd7984a89475ecfcfa12b92bc81/scripts/publisher.py#L1
  ```bash
  ubuntu@payloadpi:~$ python
  
  Command 'python' not found, did you mean:
  
    command 'python3' from deb python3
    command 'python' from deb python-is-python3
  ```
- ROS is no longer supporting Python2. http://wiki.ros.org/UsingPython3#:~:text=Python%202.7%20will%20no%20longer,Python%202%20to%20Python%203.
- DVL can produce more frequent messages than 10hz, or lower than 10hz.
  https://github.com/waterlinked/dvl-a50-ros-driver/blob/4e8f887a061cdcd7984a89475ecfcfa12b92bc81/scripts/publisher.py#L58
  ```
  ubuntu@payloadpi:~/ros_ws$ rostopic hz /transducer_report
  subscribed to [/transducer_report]                                                             
  average rate: 12.944
          min: 0.000s max: 0.128s std dev: 0.05796s window: 12
  ubuntu@payloadpi:~/ros_ws$ rostopic hz /position_report 
  subscribed to [/position_report]                                                               
  average rate: 3.536        
          min: 0.238s max: 0.367s std dev: 0.05933s window: 4
  ```

I have seperate c++ node for serial client. If needed I can merge them into one.
Because every aspect of it is different, I didn't want to send a pull request with complete new code.

**Comment by JPJSea** [comment]
It would be great if this would work for Melodic. 

At the moment there are errors, firstly because of the json library path being different on Ubuntu 18.04 (see: [https://stackoverflow.com/questions/36861355/fatal-error-with-jsoncpp-while-compiling](https://stackoverflow.com/questions/36861355/fatal-error-with-jsoncpp-while-compiling)) and secondly because the version of boost is too outdated for [`error.failed()`](https://github.com/GSO-soslab/waterlinked_dvl_ros/blob/c7828562b7ef293fcd1797e950636bd562e56947/src/waterlinked_dvl/waterlinked_dvl_tcp.cpp#L42) and [`boost::asio::io_context`](https://github.com/GSO-soslab/waterlinked_dvl_ros/blob/c7828562b7ef293fcd1797e950636bd562e56947/src/waterlinked_dvl/waterlinked_dvl_tcp.h#L15).

I tried to simply change `boost::asio::io_context` to `boost::asio::service` to just get it to build and then handled the error without `error.failed()` but then I started to experience further errors to do with the json library and had to give up.

**Comment by incebellipipo** [comment]
Let me give it a try. In the meantime can you post the the build error that cmake complains about jsoncpp's path?

**Comment by JPJSea** [comment]
I left the lab for the day however the path error was simply:
`fatal error: json/json.h: No such file or directory`
The library was on the system at in /usr/include/jsoncpp/json but cmake couldn't find it

I just tried to recreate this in a clean Ubuntu 18.04 VM at home and I couldn't. I'm still getting the boost errors obviously, but CMake now has no trouble finding the json library so I'm not sure if you'll be able to recreate this. I'll take another look at it tomorrow when I'm back in the lab. It could be something messed up with that particular Linux box.

Also I made a typo before when I wrote this:
> I tried to simply change `boost::asio::io_context` to `boost::asio::service` 

It should say `boost::asio::io_service'

**Comment by incebellipipo** [comment]
[This](https://github.com/GSO-soslab/waterlinked_dvl_ros/commit/c7828562b7ef293fcd1797e950636bd562e56947) sould fix it.

Install your dependencies with the following command:
```bash
rosdep install --from-paths src --ignore-src --rosdistro=melodic -y
```
If it doesn't work install the `libjsoncpp-dev` package using `apt`

For future references, comments such as https://github.com/waterlinked/dvl-a50-ros-driver/issues/9#issuecomment-948775266 should be an issue in this https://github.com/GSO-soslab/waterlinked_dvl_ros repository since I'm not the maintainer of this repository. I want to discuss migrating to C++11 in this issue.
